File: sampler/slice.m
%%%% Slice sampling (1d) algorithm by Neal (2003) %%%%

% inputs:
%   q0: initial state of the parameters
%   l0: initial log-density
%   logf: log-density function of q
%   w: estimate size of a slice
%   m: integer limiting th size of a slice to mw
% outputs:
%   q: new state of the parameter following N(q;0,C)*lik
%   l: new log-density


function [q,l] = slice(q0,l0,logf,w,m,bdy)
if ~exist('w','var')
    w=1;
end
if ~exist('m','var')
    m=5;
end
if ~exist('bdy','var')
    bdy=[-Inf,+Inf];
end

% log-denisty threshhold (defines a slice)
logy = l0 + log(rand);

% stepp out to obtain the [L,R] range
L = q0 - w.*rand;
R = L + w;
J = floor(m.*rand);
K = (m-1) - J;

% make sure [L,R] is within boundary
L = max([L,bdy(1)]);
R = min([R,bdy(2)]);

while 1
    if J<=0 || logy>=logf(L)
        break
    end
    L = L - w;
    L = max([L,bdy(1)]);
    J = J - 1;
end

while 1
    if K<=0 || logy>=logf(R)
        break;
    end
    R = R + w;
    R = min([R,bdy(2)]);
    K = K - 1;
end


% skrink to obtain a sample
while 1
    q = L + rand.*(R-L);
    l = logf(q);
    if l>logy
        break;
    end
    % shrink the bracket and try a new point
    if q<q0
        L = q;
    else
        R = q;
    end
end

end

File: sampler/slice.R
# This is a 1-d slice sampling

slice <- function(logf, x0, a=-Inf, b=Inf, w = 1, m = 5){ # this is used to generate one sample
	
	logy <- logf(x0) - rexp(1)
			
	# Stepping out to obtain the [L, R] range
	u = runif(1);
	L = x0 - w*u;
	R = L + w;
	v = runif(1);
	J = floor(m*v);
	K = (m-1) - J;
	
	# make sure [L, R] is within [a, b]
	L <- max(L, a)
	R <- min(R, b)
	
	
	while (J>0 && logy< logf(L)) {
		L = L - w;
		L <- max(L, a)
		J = J - 1;
	}
	#browser()
	
	while (K>0 && logy<logf(R)) {
		R = R+w;
		R <- min(R, b)
		K = K-1;
	}
	
	# Shrinkage to obtain a sample
	u = runif(1);
	x1 <- L + u*(R-L);
	while (logy > logf(x1)) {
		if (x1 < x0) L <- x1
		else R <- x1
		
		u = runif(1);
		x1 = L + u*(R-L);
	}
	
	return(x1)
}
